Sources/NIO/EventLoop.swift

```diff
@@ -1227,41 +1292,47 @@ public final class MultiThreadedEventLoopGroup: EventLoopGroup {
             return
         }
 
-        var error: Error? = nil
+        var result: Result<Void, Error> = .success(())
 
         for loop in self.eventLoops {
             g.enter()
-            loop.closeGently().recover { err in
-                q.sync { error = err }
-            }.whenComplete { (_: Result<Void, Error>) in
+            loop.initiateClose(queue: q) { closeResult in
+                switch closeResult {
+                case .success:
+                    ()
+                case .failure(let error):
+                    result = .failure(error)
+                }
                 g.leave()
             }
         }
 
         g.notify(queue: q) {
-            let failure = self.eventLoops.map { try? $0.close0() }.filter { $0 == nil }.count > 0
-
-            // TODO: For Swift NIO 2.0 we should join in the threads used by the EventLoop before invoking the callback
-            //       to ensure they're really gone (#581).
-            if failure {
-                error = EventLoopError.shutdownFailed
-            }
-
-            handler(error)
-
-            let callbacks: [(DispatchQueue, (Error?) -> Void)] = self.shutdownLock.withLock {
-                guard case .closing(let callbacks) = self.runState else {
-                    fatalError("runState was \(self.runState), expected .closing")
-                }
-
-                self.runState = .closed(error)
-
-                return callbacks
+            for loop in self.eventLoops {
+                loop.syncFinaliseClose()
             }
+            self.shutdownLock.withLock {
```

Comment:
Are we deliberately changing the scope of this lock?

Reply:
great catch, we're definitely not. fixing